org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceSource.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) OSGi Alliance (2019, 2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.osgi.test.junit5.service;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+@Target({
+	ElementType.ANNOTATION_TYPE, ElementType.METHOD
+})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@ArgumentsSource(ServiceArgumentsProvider.class)
+public @interface ServiceSource {
+
+	static long DEFAULT_TIMEOUT = 200l;
+
+	Class<?> serviceType() default Object.class;
+
+	/**
+	 * Filter string used to target more specific services using the
+	 * {@code String.format} pattern. Must use valid OSGi filter syntax.
+	 *
+	 * @return The filter string.
+	 */
+	String filter() default "";
+
+	/**
+	 * Optional arguments to the format string provided by {@link #filter()}.
```

Comment:
How is this to be used? It is not clear from the doc here what this looks like in use. Remember, people will read this doc to see how to use the annotation. I assume that these are passed to the `filter()` format string to be formatted? Since these have to be compile time constants, why would someone not just put them into the `filter()` string directly?